Matter_Engine/Create/MaterialComposition.cs

```diff
@@ -0,0 +1,130 @@
+        public static MaterialComposition CombinedMaterialComposition(IEnumerable<IElementM> elements)
+        {
+            return MaterialComposition(elements.Select(x => x.IMaterialComposition()), elements.Select(x => x.ISolidVolume()));
+        }
+
+        /***************************************************/
+
+        [Description("Creates a MaterialComposition for a collection of Materials given their ratios")]
+        [Input("materials", "The Materials the MaterialCombination is composed of")]
+        [Input("ratios", "The ratios of each material, the number of ratios must match the number of materials", typeof(Ratio))]
+        [Output("materialComposition", "A material composition composed of the provided materials and ratios scaled so that their sum equals 1")]
+        public static MaterialComposition MaterialComposition(IEnumerable<Material> materials, IEnumerable<double> ratios)
+        {
+            if (Math.Abs(1 - ratios.Sum()) > Tolerance.Distance)
+            {
+                Engine.Reflection.Compute.RecordWarning("The ratios sum did not equal 1, the values have been factored to do so");
+                double factor = 1 / ratios.Sum();
+                return new MaterialComposition(materials, ratios.Select(x => x * factor));
+            }
+
+            return new MaterialComposition(materials, ratios);
+        }
+
+        /***************************************************/
+
+        [Description("Creates a MaterialComposition for a collection of MaterialCompositions given their ratios")]
+        [Input("materials", "The MaterialCompositions the MaterialCombination is composed of")]
+        [Input("ratios", "The ratios of each material, the number of ratios must match the number of MaterialCompositions", typeof(Ratio))]
+        [Output("materialComposition", "A material composition composed of the provided MaterialComposition materials and ratios factoring both the inputted ones and the ones in the existing MaterialCompositions")]
+        public static MaterialComposition MaterialComposition(IEnumerable<MaterialComposition> materialCompositions, IEnumerable<double> ratios)
+        {
+            List<Material> allMaterials = new List<Material>();
+            List<double> allRatios = new List<double>();
+
+            List<MaterialComposition> localMatComps = materialCompositions.ToList();
+            List<double> localRatios = ratios.ToList();
+
+            if (localMatComps.Count != localRatios.Count)
+                return null;
+
+            for (int j = 0; j < localMatComps.Count; j++)
+            {
+                for (int i = 0; i < localMatComps[j].Materials.Count; i++)
+                {
+                    bool existed = false;
+                    for (int k = 0; k < allMaterials.Count; k++)
+                    {
+                        if (allMaterials[k].Equals(localMatComps[j].Materials[i]))
```

Comment:
How should one do an equality check for materials, this does not work, presumably as it's to strict.
Would it be enough to compare: Materials Name, MaterialProperties Name and MaterialProperties Types?
Or does this warrant something else?

@IsakNaslundBh  @al-fisher @FraserGreenroyd

Reply:
For all structural adapters, same name is assumed to be same material. If that is not deemed safe enough here, we could of course compare them differently.

Agreed that reference equals, as is currently done, probably is a bit to strict, so name/checking all properties/something similar I think would be better.

Could try to use a similar methodology here as for the adapters, through EqualityComparers. We have a few, quite simple ones, already implemented in BHoM_Engine (for name/guid for example) and some in Structure_Engine for node distance checks etc. This to ensure that whatever we do here could be used in other places as well.